docs/RFCS/00000000_simple_common_table_expressions.md

```diff
@@ -0,0 +1,280 @@
+       project
+  WHERE
+        users.id = tasks.user_id
+        projects.title = tasks.project_id
+  GROUP BY
+           users.email,
+           projects.title
+),
+
+--- Calculates the total tasks per each project
+total_tasks_per_project AS (
+  SELECT
+         project_id,
+         count(*) as task_count
+  FROM tasks
+  GROUP BY project_id
+),
+
+--- Calculates the projects per each user
+tasks_per_project_per_user AS (
+  SELECT
+         user_id,
+         project_id,
+         count(*) as task_count
+  FROM tasks
+  GROUP BY user_id, project_id
+),
+
+--- Gets user ids that have over 50% of tasks assigned
+overloaded_users AS (
+  SELECT tasks_per_project_per_user.user_id,
+
+  FROM tasks_per_project_per_user,
+       total_tasks_per_project
+  WHERE tasks_per_project_per_user.task_count > (total_tasks_per_project / 2)
+)
+
+SELECT
+       email,
+       task_list,
+       title
+FROM
+     users_tasks,
+     overloaded_users
+WHERE
+      users_tasks.user_id = overloaded_users.user_id
+```
+
+This query is executable by the proposed subset since every named result set is
+used at most once. As you can see, trying to represent this massive query as a
+single nested set of subselects would be much more difficult to read. This
+improvement in readability is a major reason why CTEs are popular, and
+something that we could easily provide today without waiting for temporary
+tables.
+
+# Reference-level explanation
+
+CockroachDB already supports using arbitrary plans as plan data sources, so the
+bulk of the implementation of the proposed subset of CTEs is adding syntactic
+sugar and a name resolution mechanism for CTE subclauses.
+
+Specifically, the planner will be augmented to include a naming environment
+that is composed of one frame per CTE statement. Each frame will consist of a
+mapping from CTE clause name to a query plan for executing the statement
+corresponding to the name, as well as any column aliases the user might have
+provided. The environment will be treated as a stack. Planning a CTE pushes
+a new frame onto the stack, and finishing planning that CTE pops the frame.
+
+Data source resolution will be augmented to search through the naming
+environment from the top of the stack to the bottom before searching for
+tables. CTE names can shadow table names and other CTE names in an outer scope,
+but an individual CTE statement can't contain more than one clause with any
+given name.
+
+To enforce this proposal's restrictions, the naming environment will also
+include a flag on each named clause that is set when it is used as a data
+source in another clause or the main statement. This flag will allow the
+planner to detect when a query tries to reference a table more than once and
+return a suitable error.
+
+Because of this proposal's restrictions, temporary table infrastructure is not
+necessary as each CTE clause will stream its output to the plan that references
+it just like an ordinary CockroachDB plan tree.
```

Comment:
Just curious whether if we used external storage here if the streamed data gets too big whether this proposal will not have restrictions

Reply:
"use external storage here"  == implement temp tables